Framework/MDAlgorithms/src/AccumulateMD.cpp

```diff
@@ -0,0 +1,467 @@
+ * @returns true if the file exists
+*/
+bool fileExists(const std::string &filename) {
+  if (filename.empty())
+    return false;
+  Poco::File test_file(filename);
+  return test_file.exists();
+}
+
+/*
+ * Remove anything from input_data which is already in current_data
+ * @param input_data :: Vector of input data sources
+ * @param current_data :: Vector of data sources previously appended to
+ * workspace
+ * @param psi :: Vector of goniometer angle psi containing a value for each data
+ * source
+ * @param gl :: Vector of goniometer anglegl containing a value for each data
+ * source
+ * @param gs :: Vector of goniometer angle gs containing a value for each data
+ * source
+ * @param efix :: Vector of data source energy values in meV
+*/
+void filterToNew(std::vector<std::string> &input_data,
+                 std::vector<std::string> &current_data,
+                 std::vector<double> &psi, std::vector<double> &gl,
+                 std::vector<double> &gs, std::vector<double> &efix) {
+  for (size_t i = input_data.size(); i > 0; i--) {
+    if (appearsInCurrentData(input_data[i - 1], current_data)) {
+      input_data.erase(input_data.begin() + i - 1);
+      psi.erase(psi.begin() + i - 1);
+      gl.erase(gl.begin() + i - 1);
+      gs.erase(gs.begin() + i - 1);
+      efix.erase(efix.begin() + i - 1);
+    }
+  }
+}
+
+/*
+ * Check if the named data source is in the vector of data currently in the
+ * workspace
+ * @param data_source :: Name of a data source
+ * @param current_data :: Vector of data sources previously appended to
+ * workspace
+ * @returns true if the named data source appears in the vector of current data
+*/
+bool appearsInCurrentData(const std::string &data_source,
+                          std::vector<std::string> &current_data) {
+  for (auto reverse_iter = current_data.rbegin();
+       reverse_iter != current_data.rend(); ++reverse_iter) {
+    if (data_source == *reverse_iter) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/*
+ * Return a vector of the names of files and workspaces which have been
+ * previously added to the workspace
+ * @param ws_history :: History of the workspace
+ * @returns a vector of the names of data_sources which have previously been
+ * appended to the workspace
+*/
+std::vector<std::string>
+getHistoricalDataSources(const WorkspaceHistory &ws_history) {
+  // Using a set so we only insert unique names
+  std::set<std::string> historical_data_sources;
+
+  // Get previously added data sources from DataSources property of the original
+  // call of CreateMD and any subsequent calls of AccumulateMD
+  auto view = ws_history.createView();
+  view->unrollAll();
+  const std::vector<HistoryItem> history_items = view->getAlgorithmsList();
+  for (auto iter = history_items.begin(); iter != history_items.end(); ++iter) {
+    auto alg_history = iter->getAlgorithmHistory();
+    if (alg_history->name() == "CreateMD" ||
```

Comment:
Would be better to have CreateMD provide a static public std::string algorithm name. I'm not sure anyone would remember to update this if they renamed CreateMD. You may as well do the same for AccumulateMD.